packages/app-builder-lib/src/platformPackager.ts

```diff
@@ -504,7 +504,7 @@ export abstract class PlatformPackager<DC extends PlatformSpecificBuildOptions>
 
     const _computeFileSets = (matchers: Array<FileMatcher>) => {
       return computeFileSets(matchers, this.info.isPrepackedAppAsar ? null : transformer, this, isElectronCompile).then(async result => {
-        if (!this.info.isPrepackedAppAsar && !this.info.areNodeModulesHandledExternally) {
+        if (!this.info.isPrepackedAppAsar && !this.info.areNodeModulesHandledExternally && !this.info.config.excludeNodeModulesFromAsar) {
```

Comment:
This is actually already handled by `beforeBuild` logic (albeit only if the jsdocs are read by the end-user/dev). Should just need to return `false` from `beforeBuild` hook
https://github.com/electron-userland/electron-builder/blob/9bca417a8c27facb289dac2cfcbb4ef1c111090f/packages/app-builder-lib/src/packager.ts#L525-L533

Reply:
While writing this MR, I also found that returning `false` in `beforeBuild` can achieve this functionality. However, the literal meaning of `beforeBuild` is really confusing. It’s not clear that it can be used to exclude `node_modules` from being packed into the ASAR file.

Reply:
Yeahh, that was a patch fix to avoid adding yet another configuration property. I tried to make it clear in the docs, but docs aren't always read unfortunately.

Related, for next major semver, I'm thinking it may be worth moving all native module handling (`npmRebuild`, `nativeRebuilder`, `buildDependenciesFromSource`, etc.) to nest into a single configuration object. What are your thoughts? I feel like our electron-builder-config.js is getting really nasty/complex. Could also be an optimal time to remove `nativeRebuilder` too

Reply:
> Related, for next major semver, I'm thinking it may be worth moving all native module handling (npmRebuild, nativeRebuilder, buildDependenciesFromSource, etc.) to nest into a single configuration object. What are your thoughts? I feel like our electron-builder-config.js is getting really nasty/complex. Could also be an optimal time to remove nativeRebuilder too

Indeed, these native module configurations should be removed. Some of them are redundant and overly complex, and many people don't use them.
Moreover, if we can eventually remove `app-builder-bin` and rewrite everything in TypeScript, we could achieve better cross-platform support, such as for the Loongson architecture.

Reply:
>If we have the node module collector fully working and since it's not a performance hog, I don't think it's necessary to implement this request. app-builder-bin always collected node modules before filtering, so I don't see any logic that was "broken" during the Golang->JS migration. Curious on your thoughts

I agree with your opinion. We can provide a detailed explanation in the docs about the functionality of `beforeBuild` for excluding `node_modules` from being packed into the ASAR file.

Reply:
>Moreover, if we can eventually remove app-builder-bin and rewrite everything in TypeScript

I absolutely love this goal. It's going to take a hot minute to get there though in terms of x-platform support for things like icon generation, appimages, rcedit, and wine 😬